internal/controllers/cluster/cluster_controller.go

```diff
@@ -215,6 +219,22 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster)
 func (r *Reconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster) (reconcile.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
 
+	// Call the BeforeClusterDelete hook to before proceeding further.
+	if feature.Gates.Enabled(feature.RuntimeSDK) {
+		hookRequest := &runtimehooksv1.BeforeClusterDeleteRequest{
+			Cluster: *cluster,
+		}
+		hookResponse := &runtimehooksv1.BeforeClusterDeleteResponse{}
+		if err := r.RuntimeClient.CallAllExtensions(ctx, runtimehooksv1.BeforeClusterDelete, hookRequest, hookResponse); err != nil {
```

Comment:
I know I said it should be fine, but thinking about it some more. That's our current description the proposal:

> This hook is called after the Cluster has been deleted by the user, and immediately before objects existing in the Cluster are going to be deleted. Runtime Extension implementers can use this hook to execute cleanup tasks for the add-ons and block deletion of the Cluster and descendant objects until everything is ready. 

That would simply be not true. It wouldn't actually be a BeforeClusterDelete it's more like a "before and during" delete hook. If we take it literally it's still "before **cluster**", but we definitely have to change our description (probably our godoc too).

I would prefer if it would be just called until it succeeds once as that's probably what a user would expect (even from the hook name alone, if they know that a cluster delete includes all cluster resources). 

I know this would mean we move this part out of this PR to the one where we introduce the hooks which need tracking.

Opinions?
(cc @fabriziopandini @killianmuldoon @chrischdi)

Reply:
Yup we already have some ideas about tracking / early prototype. Essentially we would set an annotation if we detect in that a hook should be called. We then remove it once the hook(s) were called successfully and move on. move on has to result in a change which makes sure that we don't detect again that the hook should be called. But I would defer a more detailed discussion to somewhere else / the next PR. (just to not get sidetracked on this conversation)